### PR TITLE
prioritize plain R code over knitr

### LINKED ISSUE #743: R files generating knitr templates are parsed as Rmd
If I have a file that generates `knitr` templates, e.g.

    writeLines('
    CreateTemplate <- function(x) {
    return(sprintf("
    ```{r code}
    foo <- function(x) x+%d
    foo(5)
    ```", x))
    }', tmp <- tempfile())

`parse(tmp)` works just fine, however, `lintr` assumes this is an Rmd based on matching some chunk regex:

```
lintr::lint(tmp)
# /tmp/RtmpN8X7Dj/file3d0536f05e5d:4:1: error: Missing chunk end for chunk (maybe starting at line 4).
# ```{r code}
# ^
```

Should we try parsing the file as R code first, then checking Rmd chunks afterwards?

### LINKED ISSUE #879: lintr:::get_source_expressions fails when a code comment contains an r script command surrounded by backticks
We were surprised our files linted correctly despite having obvious linting issues, so I have investigated and narrowed it down to this minimal reprex:
```r
# `r print("7")`
function() 2<=3
```
The above doesn't lint.

I further narrowed the issue down to backticks in combination with r and something because the following do lint correctly:
```r
# `print("7")`
function() 2<=3

# `r`
function() 2<=3

# r print("7")
function() 2<=3
```
In short - all three are needed: backticks, r and something in addition to r.

Then, I dug inside inflix_spaces_linter and tracked the issue to the parsing function, Given a file "test.R", `lintr:::get_source_expressions("test.R")` returns:
image
Without the comment, it parses correctly:
image

Environment:
lintr version 2.0.1
R: 4.1.0

Is this a known issue?

### LINKED ISSUE #1406: `extract_r_source()` is overly eager with detecting knitr patterns
It seems the test code is linted as Rmd for some reason?!
The NA should be investigated in a separate issue. Maybe we stumbled on an actual bug in the extractor code.

_Originally posted by @AshesITR in https://github.com/r-lib/lintr/pull/1403#discussion_r900382755_

### PR DESCRIPTION
**Pull request by AshesITR**
fixes #1406 
fixes #743 
fixes #879

### CODE REVIEW

File: R/comment_linters.R [hunk 1]
```diff
@@ -79,7 +79,7 @@ commented_code_linter <- function() {
 
 # is given text parsable
 parsable <- function(x) {
-  if (is.na(x)) {
+  if (anyNA(x)) {
```
Comment (AshesITR) [inline R/comment_linters.R:82]: This is required so that `parsable()` is vectorized over `x`.

**Comment by MichaelChirico** [comment]
does this fix #743

**Comment by AshesITR** [comment]
yes.

**Comment by MichaelChirico** [comment]
#879 too?

(sorry I knew there were two but search was eluding me)

File: NEWS.md [hunk 2]
```diff
@@ -7,6 +7,7 @@
 
 ## Bug fixes
 
+* `get_source_expressions()` no longer fails on R files that match a knitr pattern (#1406, @AshesITR).
```
Comment (MichaelChirico) [inline NEWS.md:10]: please add the other issue #s here too

File: tests/testthat/test-get_source_expressions.R [hunk 3]
```diff
@@ -249,6 +249,13 @@ test_that("#1262: xml_parsed_content gets returned as missing even if there's no
   expect_identical(source_expressions$expressions[[1L]]$full_xml_parsed_content, xml2::xml_missing())
 })
```
Comment (MichaelChirico) [inline tests/testthat/test-get_source_expressions.R:251]: nit: i usually like to add tests from duplicate issues just to ensure future updates don't regress one issue but not the others, i.e. just copy-paste the reported reprex with expected output. i think we're pretty confident that the issues are the exact same root cause, but better safe than sorry.

**Review by MichaelChirico — APPROVED** [review]
